File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int zeros = 0;
        int productwithout = 1;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] == 0) {
                ++zeros;
            } else {
                productwithout *= nums[i];
            }
        }

        if (zeros > 1) {
            vector<int> ret(nums.size(), 0);
            return ret;
        }

        vector<int> ret;
        for (int i = 0; i < nums.size(); ++i) {
            if (zeros == 1) {
                if (nums[i] == 0) {
                    ret.push_back(productwithout);
                } else {
                    ret.push_back(0);
                }
            } else {
                ret.push_back(productwithout/nums[i]);
            }
        }

        return ret;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        answer = [0] * len(nums)
        answer[0] = 1
        for i in range (1, len(nums)):
            answer[i] = nums[i - 1] * answer[i - 1]
        
        right = 1
        for i in reversed(range(len(nums))):
            answer[i] = answer[i] * right
            right *= nums[i]

        return answer